build_barback/lib/src/transformer/transformer.dart

```diff
@@ -137,15 +138,13 @@ class _TransformAssetWriter implements AssetWriter {
   _TransformAssetWriter(this.transform);
 
   @override
-  Future writeAsBytes(build.AssetId id, List<int> bytes) {
-    transform.addOutput(barbackAssetFromBytes(id, bytes));
-    return new Future.value(null);
+  Future writeAsBytes(build.AssetId id, FutureOr<List<int>> bytes) async {
```

Comment:
The `.complete()` in `BuildStep` will wait for this so I don't think it matters here.

I'm also considering not changing `AssetWriter` and only changing the interface in `BuildStep`. That way most of our implementations won't need to care about it and there are only a few packages which see those directly anyway.